proctor-common/src/main/java/com/indeed/proctor/common/ForceGroupsOptionsStrings.java

```diff
@@ -17,60 +23,206 @@
  *     <li>It's a concatenation of string elements separated by commas. (e.g., my_btn_tst1,default_to_fallback)</li>
  *     <li>Each element represents a forced test group, or an option.</li>
  *     <li>A forced group is specified by a test name followed by a bucket value (e.g., my_btn_tst1)</li>
+ *     <li>A forced payload is specified by a force group followed by a semicolon and a payload definition (e.g., my_btn_tst1;doubleValue:0.2) </li>
  *     <li>A option is specified by predefined tokens that doesn't contain integers (e.g., default_to_fallback) </li>
  *     <li>If two elements conflict (e.g., specifying different buckets for the same test),
  *     the latter takes precedence</li>
  * </ul>
  */
 public class ForceGroupsOptionsStrings {
     private static final Logger LOGGER = LogManager.getLogger(ForceGroupsOptionsStrings.class);
+    private static final int FORCE_PARAMETER_BUCKET_IDX = 0;
+    private static final int FORCE_PARAMETER_PAYLOAD_IDX = 1;
+    private static final int FORCE_PARAMETER_MAX_SIZE = 2;
 
     private ForceGroupsOptionsStrings() {
     }
 
     @Nonnull
-    public static ForceGroupsOptions parseForceGroupsString(@Nullable final String forceGroupsString) {
+    public static ForceGroupsOptions parseForceGroupsString(@Nullable final String forceGroupsString, final Set<String> forcePayloadTests) {
         final ForceGroupsOptions.Builder builder = ForceGroupsOptions.builder();
         if (forceGroupsString == null) {
             return builder.build();
         }
-        // using single char in split regex avoids Pattern creation since java8
-        final String[] pieces = forceGroupsString.split(",");
-        // detect integer number from end of string
-        for (final String rawPiece : pieces) {
-            final String piece = rawPiece.trim();
-            if (piece.isEmpty()) {
-                continue;
+
+        int startIndex = 0, numQuotes = 0;
+        final List<Character> brackets = new ArrayList<>();
+
+        // iterate through string to find force groups and force payloads
+        for (int forceGroupIndex = 0; forceGroupIndex < forceGroupsString.length(); forceGroupIndex++) {
+            if (forceGroupsString.charAt(forceGroupIndex) == '[' || forceGroupsString.charAt(forceGroupIndex) == '{') {
+                brackets.add(forceGroupsString.charAt(forceGroupIndex));
+            } else if (forceGroupsString.charAt(forceGroupIndex) == ']' || forceGroupsString.charAt(forceGroupIndex) == '}') {
+                brackets.remove(brackets.size()-1);
+            } if (forceGroupsString.charAt(forceGroupIndex) == '"') {
```

Comment:
I'm afraid this is not a perfect solution and it will break with some edge cases. For example,
`abc1;stringValue:"te\"st",def2;stringValue:"test2"` (Java string literal: `"abc1;stringValue:\"te\\\"st\",def2;stringValue:\"test2\""`) seems not able to be parsed with this code.

However, I don't have any good suggestion right now unlike `ObjectMapper` case. This sounds like a valid parsing problem we have to implement by ourselves.

Reply:
Enforcing some format sounds a good approach but I'm not sure what format could help with the parsing... If force groups comes from only URL parameters, we could enforce using multiple URL parameter like `?forceGroups=abc1;stringValue:"test"&forceGroups=mytest2&forceGroups=abtest0` but that seems not possible.

Otherwise, we need more complex logic to parse all possible edge cases correctly (including ignoring quotes that preceded by `\`), or give up supporting some types of inputs.

Regarding edge cases, I can come up with a few more cases that may or may not work with the current implementation.
* a string value including `"` (e.g., `abc1;stringValue:"te\"st"`)
* a string value including `\"` (e.g., `abc1;stringValue:"te\\\"st"`)
* an array of strings including `[`  or `]` (e.g., `abc1:stringArray:["[1]", "[2]", "]"]`)

Reply:
Alternative idea is to create a separate parameter for forcePayloads (like `?prforceGroups=abc1,mytest2&prforcePayloads=abc:stringValue:"test"&prforcePayloads=mytest:stringValue:"foo"`) but that probably require a lot of changes from the current design.

Reply:
I have been working on playing around with regex to implement parsing. It would be a little more greedy complexity wise but I so far have come up with this that might work: https://regex101.com/r/Q1KOXS/1

Reply:
I went ahead and made the changes to use the regular expression in the link above and added more unit tests with some of your edge cases suggest above.

Reply:
Thanks! Using regular expression sounds a good idea and comments explaining it is nice. I will try to understand the regex and review if this works for any other edge cases. (My guess is this is practically good enough even if it won't work with some artificial edge cases)

Reply:
I have reviewed regex and I think we have at least two cases that cannot be parsed with the current method.

* `stringValue:"\\"` (a string with a single character `\`). I think this won't match with the regex due to the pattern to ignore `\"`
*  `stringArray:["]a"]`. I think regex will match with a wrong part `stringArray:["]`

Reply:
This may be a overkill but one of the part of the regex (after the payload type part such as `stringValue:`) is basically matching with (a subset of) a valid JSON so we could utilize some common solution like https://stackoverflow.com/questions/32155133/regex-to-match-a-json-string

Reply:
> * `stringValue:"\\"` (a string with a single character `\`). I think the regex won't match with this due to the pattern to ignore `\"`
I had to add a new case to handle this. \".*?\\\\+?\" quote followed by any character then 2 or more backslashes followed by a quote. https://regex101.com/r/x5t2rN/1


> * `stringArray:["]a"]`. I think regex will match with a wrong part `stringArray:["]`
 I was able to edit the current regex by adding (?=,|$) to ignore brackets that are not followed by a comma or end of string.

Reply:
I know this won't likely happen in real usecase but I see some edge cases 
* a string with several `\\` and `\"`. e.g., `abc1;stringValue:"\"\\\"\\\""`
* a string array with `],`. e.g, ` abc2;stringArray:["],"]`.

Using the service linked from your URL, I tried to construct more comprehensive regex (using ideas from the regex to validate JSON) and here's a link https://regex101.com/r/wdvEnD/1

Reply:
the main updates from your regex are

* using `\"(\\.|[^\"\\]+)*\"` for matching with string. 
* using `[(<string or number>(, <string or number>)*)?]` for matching with array.

Reply:
Updated regex and documentation in comments.

Reply:
Thanks. Looks good to me!